python/lsst/daf/butler/core/connectionString.py

```diff
@@ -0,0 +1,168 @@
+    2. RegistryConfig for explicitly defined keys
+    3. and DbAuth instance with which it was created.
+
+    The 'db' string must be a connection URL, i.e. must start with a database
+    dialect and contain, at least, the `database` component or the host name.
+
+    Components `username`, `password`, `host`, `port` or `database` can also be
+    specified explicitly as keys in the config when not part of the `db` string
+    in Config.
+    When `username` and `password` are not part of the `db` string or stated
+    explicitly in Config they will be read from `~/.lsst/db-auth.paf` file by
+    matching matching `host` and `port` components of provided URL.
+
+    Parameters
+    ----------
+    dbAuth : `DbAuth`
+        DbAuth instance to use in order to resolve username and passwords.
+    conUrl : `sqlalchemy.engine.URL`
+        URL instance from which connection string will be created
+    """
+    def __init__(self, dbAuth, conUrl):
+        super().__init__(conUrl.drivername, **conUrl.translate_connect_args())
+        self._dbAuth = dbAuth
+
+    @property
+    def dialect(self):
+        """Returns the dialect part of the connection string."""
+        return self.get_backend_name()
+
+    @property
+    def driver(self):
+        """Returns the driver part of the connection string."""
+        return self.get_driver_name()
+
+    def checkDialectDriverExist(self):
+        """Verifies that the targeted dialect and driver exist and are
+        installed.
+
+        Attempts to load the correct dialect and driver from SQLAlchemy. If
+        load is successfull, returns True, otherwise returns False.
+        """
+        try:
+            self._conStr.get_dialect()
+        except NoSuchModuleError:
+            return False
+        return True
+
+    def asDict(self):
+        """Returns connection string arguments as a dictionary."""
+        return self._conStr.translate_connect_args()
+
+
+class ConnectionStringFactory:
+    """Factory for `ConnectionString` instances.
+
+    Parameters
+    ----------
+    path : `str` or None, optional
+        Path to configuration file. Defaults to '~/.lsst/daf-auth.yaml'.
+    envVar : `str` or None, optional
+        Name of environment variable pointing to configuration file. Defaults
+        to "LSST_DB_AUTH"
+    authList : `list` [`dict`] or None, optional
+        Authentication configuration.
+
+    Notes
+    -----
+    At least one of ``path``, ``envVar``, or ``authList`` must be provided;
+    generally ``path`` should be provided as a default location.
+    """
+
+    keys = ('username', 'password', 'host', 'port', 'database')
+
+    def __init__(self, path=None, envVar=None, authList=None):
+        if path is None:
+            path = DB_AUTH_PATH
+        if envVar is None:
+            envVar = DB_AUTH_ENVVAR
+        self._dbAuth = DbAuth(path, envVar, authList)
+
+    def __str__(self):
+        return f"{self.__class__.__name__}@{self._dbAuth.authList}"
```

Comment:
This is dangerous; we don't want to accidentally expose the user's auth information in a log message or print output.  I'd remove it.